functorch/experimental/_cond.py

```diff
@@ -56,12 +56,35 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
 
     flat_true_outs, _ = pytree.tree_flatten(true_outs)
     flat_false_outs, _ = pytree.tree_flatten(false_outs)
-    assert(len(flat_true_outs) == len(flat_false_outs))
+    if len(flat_true_outs) != len(flat_false_outs):
+        raise UserError(
```

Comment:
I am not sure if it is a good idea to use UserError here as it is kind of TorchDynamo specific thing. Should we consider moving UserError to more higher level place? cc @yanboliang @anijain2305 @avikchaudhuri

Reply:
Maybe you can just throw different type of Error subclass here and use that to raise UserError in dynamo side?

Reply:
I do think this is a user error: outputs need to have the same types in both branches, which is a standard thing to require in other settings as well. 

I do worry about UserError / UserErrorType becoming a hot mess over time, though. E.g., COND_OP_RESTRICTION seems too low-level as a classifier. Something like CONTROL_FLOW_RESTRICTION would be better to cover more.

Reply:
> I do think this is a user error: outputs need to have the same types in both branches, which is a standard thing to require in other settings as well.

The question is do we expect users to take actions when seeing this error. Isn't still user who write these branches? Currently I revert it back to assert as you guys think it's too obvious to point out that outputs need to have the same length and meta in both branches.

Reply:
To be clear, I said I think these *should* be user errors, because I think the user is in control of metadata of the outputs; OTOH I guess @tugsbayasgalan is saying that exposing internal data structures to users to assert their equality across branches is not good, which I also agree with, so we just have to check equality of each part separately to describe the problem in terms of what the user can understand.

Reply:
Oh, I read it as it should not! My bad.

Reply:
Addressed